sparta/sparta/statistics/BasicHistogram.hpp

```diff
@@ -0,0 +1,63 @@
+// <BasicHistogram.h> -*- C++ -*-
+
+/**
+ * \file BasicHistogram.hpp
+ * \brief A simple histogram with programmable ranges, using sparta::Counters
+ */
+#pragma once
+
+#include "sparta/statistics/Counter.hpp"
+
+/// Histogram with programmable buckets
+class BasicHistogram
+{
+public:
+    /// constructor
+    ///\param buckets one bucket will be created per value (plus one for overflow) - values must be sorted
+    BasicHistogram(sparta::StatisticSet &sset, const std::string &name, const std::string &desc, const std::vector<int64_t> &buckets)
+    : bucket_vals_(buckets)
+    {
+        // create one counter per bucket
+        for (unsigned i = 0; i < bucket_vals_.size(); ++i)
+        {
+            const int64_t v = bucket_vals_[i];
+            std::ostringstream os_name;
+            if (v < 0)
+                os_name << name << "_n" << -v; // negative
+            else
+                os_name << name << '_' << v;
+
+            std::ostringstream os_desc;
+            if (i == 0)
+            {
+                os_desc << desc << " with values less than " << v;
+            }
+            else
+            {
+                os_desc << desc << " with values greater than or equal to " << bucket_vals_[i-1] << " and less than " << v;
+            }
+
+            ctrs_.push_back(sparta::Counter(&sset, os_name.str(), os_desc.str(), sparta::Counter::COUNT_NORMAL));
+        }
+
+        // create a final bucket for everything beyond the last value
+        std::ostringstream os_name;
+        os_name << name << "_of";
+        std::ostringstream os_desc;
+        os_desc << desc << " overflow";
+        ctrs_.push_back(sparta::Counter(&sset, os_name.str(), os_desc.str(), sparta::Counter::COUNT_NORMAL));
+    }
+
+    /// increment the bucket corresponding to 'val'
+    void addValue(int64_t val)
```

Comment:
What happens if I pass in `-1`?

Reply:
You can potentially define negative buckets, if so, it will be sorted into the appropriate bucket. Otherwise, it goes into the first bucket.

Reply:
So, if I pass in `-1` to a group of buckets 0-100, 100-200, 200-300, you'd expect it to be charged to bucket 0-100?  Does that seem right?

Reply:
I don't see how the class can know what are the expectations of the input. If -1 is invalid, there should be an assert in the user code. If negative numbers want to be binned separately, there should be a leading bucket of -1...

Reply:
The point of a framework is to not force the user to perform his/her own error checking, but to ensure the user can find issues quickly if they do something wrong.  Clearly, -1 (as well as something like 500) is not a valid input and I think this class should ensure that the values given to it should cleanly fall into a bucket that it has created (i.e. assert).  Either that, or enhance the class to allow for "out of range" values and create buckets for "beyond lower" and "beyond upper."  You have the list of valid ranges, it's not hard to derive what the upper/lower limits are.